refactor(devtools): Add support for signals.

JeanMeche opened the pull request:
The devtools now support signals.
Writable signals of primitives are editable.
Object Signal and other non-writable signals (like computed) are not editable.

![Screenshot 2023-11-29 at 23 05 49](https://github.com/angular/angular/assets/1300985/12ca4ac7-2061-44ff-947d-bd8d7a0546ad)


This a collaborative work with @ducin

JeanMeche commented on devtools/projects/ng-devtools-backend/src/lib/utils.ts:
```diff
@@ -34,3 +34,13 @@ export const isCustomElement = (node: Node) => {
+
+
+export function isSignal(prop: unknown): boolean {
+  if (typeof prop == 'function') {
```
This is open to improvements/suggestions.

JeanMeche replied:
This will likely be improved by #53268

JeanMeche replied:
Actualy, @AleksanderBodurri made a better suggestion. Let's expose `isSignal()` globally.

crisbeto commented on packages/core/src/render3/util/global_utils.ts:
```diff
@@ -65,6 +66,7 @@ export function publishDefaultGlobalUtils() {
     publishGlobalUtil('getRootComponents', getRootComponents);
     publishGlobalUtil('getDirectives', getDirectives);
     publishGlobalUtil('applyChanges', applyChanges);
+    publishGlobalUtil('isSignal', isSignal);
```
Do we have the ability to import from `@angular/core` in the extension or do we need to go through global functions for everything? Ideally we could just import this.

alxhub replied:
The extension needs to access utilities through `window` - it's not bundled with the application and can't import from application modules.

JeanMeche committed:
refactor(devtools): Add support for signals.

The devtools now support signals.
Writable signals of primitives are editable.
Object Signal and other non-writable signals (like computed) are not editable.

Co-authored-by: Tomasz Ducin <tomasz.ducin@gmail.com>

AleksanderBodurri commented on devtools/projects/protocol/src/lib/messages.ts:
```diff
@@ -67,6 +67,8 @@ export interface InjectedService {
   providedIn: Injector;
 }
 
+export type ContainerType = 'WritableSignal'|'ReadonlySignal'|false;
```
nit: `null` feels a bit more natural to me as a container type instead of `false here`

AleksanderBodurri commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/directive-explorer/property-tab/property-view/property-editor.component.html:
```diff
@@ -3,9 +3,14 @@
-            {{ valueToSubmit }}
-          </span>
+            @switch(containerType) {
+              @case('WritableSignal') { 
```
Do we care about `ReadonlySignal` as a case here?

JeanMeche replied:
No as the ReadonlySignai is not editation (we're in the editor here !)

JeanMeche commented:
@AleksanderBodurri CI is green

AleksanderBodurri reviewed (approved):
Awesome work, thank you @JeanMeche and @ducin 🙏

dylhunn commented:
This PR was merged into the repository by commit e2272750878e1f74e59121e76a01e910ad1ac607.
